### ISSUE TITLE
Instead of video title, send video ID to Matomo Media Analytics Plugin

### ISSUE THREAD

**Issue by snoesberger** [polimediaupv/paella-user-tracking#8]
@miesgre as discussed at the Opencast summit in Zaragoza, here my wish for the Matomo integration.

In the case of the Matomo Media Analytics plugin, the data sent to Matomo includes the title of the video. As the video title is not unique in Opencast, this can cause statistics to be aggregated if several videos have the same name. If we would send the video ID, we could make sure that we collect the statistics for each unique video separately. This would also make it easier to read the statistics for a single unique video from Matomo.

One could also imagine making it configurable whether the video title, ID or other data can be sent.

If you need any further information, please do not hesitate to contact me.
 
Thanks for having a look at it!

**Comment by snoesberger** [comment]
To clarify, the title is currently only sent with Paella 6, and is completely missing with Paella 7. See this comment for more information: https://github.com/polimediaupv/paella-user-tracking/issues/5#issuecomment-2039294591

The missing title information was also once a problem in Paella 6 and was fixed with this commit: https://github.com/polimediaupv/paella/commit/2a65b3ea2485606adb3010b78fa759917d2e8a79

**Comment by miesgre** [comment]
Thanks @snoesberger for the info.

Matomo documentation: [Enriching and customising media data tracking](https://developer.matomo.org/guides/media-analytics/options)

**Comment by luniki** [comment]
I'll take a look at this.

**Comment by luniki** [comment]
Originally I thought I could change this directly in this project. However, to send the video ID instead of the document title requires changes to the code using `paella-core` instead (as @miesgre linked above).

**Comment by miesgre** [comment]
I don't have the Media Analitycs plugin and I can not test it, but. **Is the title currently being sent?**

According to the documentation:

> Media Analytics will detect the media title by searching for the following pieces of information:
>
> - firstly, the data-matomo-title HTML attribute.
> - next, the data-piwik-title HTML attribute.
> - Media title from YouTube / Vimeo player.
> - title HTML attribute.
> - finally, the alt HTML attribute.

If the title is currently being sent, you may only need to change the title of the document.
maybe a
```js
document.title = `${videoId} - ${videoTitle}`;
```
or making it configurable in the config.

**Comment by snoesberger** [comment]
> I don't have the Media Analitycs plugin and I can not test it, but. **Is the title currently being sent?**

We have the Media Analytics plugin installed and at the moment the title information isn't send. The `ma_ti` parameter is always empty, other parameters like `ma_id`, `ma_pn`, etc. are set.

**Comment by luniki** [comment]
Thank you for your links @miesgre !

I have created a solution for the Opencast integration of the paella player by extending `MatomoUserTrackingDataPlugin` and this `load` method:

```
  load() {
    bindEvent(this.player, Events.STREAM_LOADED, () => {
      this.player.containerElement.querySelector('video').dataset.matomoTitle = <video-id>;
    });
    super.load();
  }
```

So essentially I am searching for the `video` element inside the player's `containerElement` to set the `data-matomo-title` attribute.

Is there a better way to access the `video` element in order to add the data attribute?

(I tried to understand the `paella-core` documentation for the correct way but failed :-( )

**Comment by miesgre** [comment]
I think right now it's the best we can do.

The only drawback I see is if at some point the `<video>` element is replaced. In that case the `data-matomo-title` attribute would be lost (I'm not sure if the STREAM_LOADED event would be thrown again in these cases).
But this should not happen in paella-core, so in theory it should work correctly

Please make a PR with the changes.
Thank you very much for your contributions.

**Comment by miesgre** [comment]
As I said, I don't have the plugin and I can't test it...
I have a question. Paella plays multiple videos... so is the plugin counting the views correctly or is it counting for each of the videos (presenter+presentation)?
Is Matomo counting twice? Maybe we should only activate monitoring of the "master" video?

**Comment by snoesberger** [comment]
I have tested @luniki's [suggested solution](https://github.com/polimediaupv/paella-user-tracking/issues/8#issuecomment-2210669100) on my local Opencast installation with the Media Analytics plugin and it works fine with single videos. But with multiple videos, no statistics are collected for Media Analytics.
In the case of multiple videos in the Paella player, the Media Analytics stats for each video are sent as form data in a single post request to Matomo, whereas in the case of a single video, the data is sent as url parameters. It seems that Matomo doesn't take into account the data sent as form data (bug in Matomo?).

I played around a bit and found a solution that works for both single and multiple videos. I also found a way to put the code directly into the pealla-user-tracking plugin instead of the Opencast integration (`MatomoUserTrackingDataPlugin`). This will make it easier to integrate with other tools like Tobira or LMSs. I'll be submitting a PR with my suggestions later.
